### PR TITLE
feat(chips): Add input chips

### PR DESCRIPTION
**Pull request by bonniezhou**
Handle adding/removing chips, including entry/exit animation and adding/removing from DOM.
Add screenshot test to show how a user would implement input chips logic (which is so simple and beautiful!)

Also:
- rename base Chip (sans ripple) to `ChipBase` and use for testing
- render leading/remove icons with proper classes
- remove dead adapter code from `ChipSet`
- removed id prop, since chips don't actually need to be aware of their id (but kept example using id's in screenshot tests)

One small nick: we are waiting for `MDCChipFoundation`'s `trailingIconInteractionHandler_` and `transitionEndHandler_` handlers to be made public ([PR here](https://github.com/material-components/material-components-web/pull/3146)). I'm open to either merging this PR and filing an issue for updating against the next MDC Web release, or postponing this PR until the next release.

### CODE REVIEW

File: packages/chips/Chip.js [hunk 1]
```diff
@@ -40,49 +46,104 @@ export class Chip extends Component {
...
+      return null;
+    }
+    const className = classnames(
+      removeIcon.props.className,
+      'mdc-chip__icon',
+      'mdc-chip__icon--trailing');
+    const props = Object.assign({},
+      removeIcon.props,
+      {className},
+      {'onClick': this.handleRemoveIconClick});
```
[the remark this answers is not in the document]
Comment (bonniezhou) [reply packages/chips/Chip.js]: It calls the foundation which isn't initialized yet :(

File: packages/chips/README.md [hunk 2]
```diff
@@ -105,15 +105,73 @@ class MyFilterChips extends React.Component {
...
+        <Chip label='Tops' selected={this.isSelected(0)} handleSelect={() => this.handleSelect(0)}/>
+        <Chip label='Bottoms' selected={this.isSelected(1)} handleSelect={() => this.handleSelect(1)}/>
+        <Chip label='Shoes' selected={this.isSelected(2)} handleSelect={() => this.handleSelect(2)}/>
       </ChipSet>
     );
   }
 }
 ```
 
+### Input chips
```
[the remark this answers is not in the document]
Comment (bonniezhou) [reply packages/chips/README.md]: They're under the `###` title "Selection"

File: packages/chips/Chip.js [hunk 3]
```diff
@@ -40,49 +46,104 @@ export class Chip extends Component {
...
   handleClick = (e) => {
     if (typeof this.props.onClick === 'function') {
       this.props.onClick(e);
     }
-    this.props.handleSelect(this.props.id);
+    this.props.handleSelect();
+  }
+
+  handleRemoveIconClick = (e) => {
+    this.foundation_.trailingIconInteractionHandler_(e); // TODO: Update with public MDCChipFoundation handler.
```
[the remark this answers is not in the document]
Comment (bonniezhou) [reply packages/chips/Chip.js]: #173 ✨

File: test/screenshot/chips/index.js [hunk 4]
```diff
@@ -54,19 +55,77 @@ class ShoppingFilters extends React.Component {
...
+      nextId: id + 1,
+    });
+  }
+
+  handleKeyDown = (e) => {
+    if (!(e.key === 'Enter') || !e.target.value) {
+      return;
+    }
+    this.addChip(e.target.value);
+    e.target.value = '';
```
[the remark this answers is not in the document]
Comment (bonniezhou) [reply test/screenshot/chips/index.js]: Added a comment in the readme

File: packages/chips/Chip.js [hunk 5]
```diff
@@ -94,10 +90,10 @@ export class ChipBase extends Component {
       removeIcon.props.className,
       'mdc-chip__icon',
       'mdc-chip__icon--trailing');
-    const props = Object.assign({},
-      removeIcon.props,
-      {className},
-      {'onClick': this.handleRemoveIconClick});
+    const props = Object.assign({}, removeIcon.props, {
+      className,
+      'onClick': this.handleRemoveIconClick,
```
Comment (moog16) [inline packages/chips/Chip.js]: the quotes are unnecessary

File: packages/chips/Chip.js [hunk 6]
```diff
@@ -40,49 +46,104 @@ export class Chip extends Component {
         this.setState({classList});
       },
       hasClass: (className) => this.classes.split(' ').includes(className),
+      eventTargetHasClass: (target, className) => target.classList.contains(className),
+      getComputedStyleValue: (propertyName) => window.getComputedStyle(this.root_).getPropertyValue(propertyName),
+      setStyleProperty: (propertyName, value) => this.root_.style.setProperty(propertyName, value),
```
[the remark this answers is not in the document]
Comment (moog16) [reply packages/chips/Chip.js]: I actually think doing inline styles like seen in this blog post is more "react" like. https://blog.cloudboost.io/using-inline-styles-in-jsx-c1d03cbe6fe0. Passing it the parent div an object is better than having a a ref to the dom node and setting .style.
```
<div style={{margin: '20px', padding: '20px'}}>
```

File: test/screenshot/chips/index.js [hunk 7]
```diff
@@ -24,9 +25,9 @@ class ShirtSizes extends React.Component {
   render() {
     return (
       <ChipSet>
-        <Chip selected={this.isSelected(0)} id={0} label='Small' handleSelect={this.handleSelect}/>
-        <Chip selected={this.isSelected(1)} id={1} label='Medium' handleSelect={this.handleSelect}/>
-        <Chip selected={this.isSelected(2)} id={2} label='Large' handleSelect={this.handleSelect}/>
+        <Chip selected={this.isSelected(0)} label='Small' handleSelect={() => this.handleSelect(0)}/>
```
Comment (moog16) [inline test/screenshot/chips/index.js]: I'm seeing a warning when you delete a chip. I think its the ripple is calling setState on unmount. I think this was an issue.

screen shot 2018-07-24 at 16 45 41

**Comment by bonniezhou** [comment]
Redoing this PR in #246 to incorporate changes in MDC Web (passing chips by id)

**Pull request closed without merging**